### PR TITLE
[CORE] adding HasIntersection for line_2d_2

### PR DESCRIPTION
**Pull request by adityaghantasala**
This is to enable to use of line 2d with bin based fast point locator.

### CODE REVIEW

File: kratos/geometries/line_2d_2.h [hunk 1]
```diff
@@ -890,6 +890,57 @@ class Line2D2 : public Geometry<TPointType>
         return false;
     }
 
+    /** Test the intersection with another geometry
+     *  Test if this geometry intersects with other line_2d_2
+     *
+     * @param  rOtherGeometry Geometry to intersect with
+     * @return True if the geometries intersect, False in any other case.
+     */
+    virtual bool HasIntersection(const BaseType& rOtherGeometry) override
```
Comment (philbucher) [inline kratos/geometries/line_2d_2.h]: please remove the virtual, it is not needed and afaik gives warnings in some compilers

File: kratos/geometries/line_2d_2.h [hunk 2]
```diff
@@ -890,6 +890,57 @@ class Line2D2 : public Geometry<TPointType>
...
+
+    /** Test intersection of the geometry with a box
+     * Tests the intersection of the geometry with
+     * a 3D box defined by rLowPoint and rHighPoint
+     *
+     * @param  rLowPoint  Lower point of the box to test the intersection
+     * @param  rHighPoint Higher point of the box to test the intersection
+     * @return            True if the geometry intersects the box, False in any other case.
+     */
+    virtual bool HasIntersection(const Point& rLowPoint, const Point& rHighPoint) override
```
Comment (philbucher) [inline kratos/geometries/line_2d_2.h]: same

File: kratos/geometries/line_2d_2.h [hunk 3]
```diff
@@ -890,6 +890,57 @@ class Line2D2 : public Geometry<TPointType>
...
+        // We get the other line's points
+        const TPointType& first_point_other  = *rOtherGeometry(0); //p3
+        const TPointType& second_point_other = *rOtherGeometry(1); //p4
+
+        // parametric coordinate of intersection on current line
+        double t = ( (first_point[0]-first_point_other[0])*(first_point_other[1] - second_point_other[1]) - (first_point[1]-first_point_other[1])*(first_point_other[0]-second_point_other[0]) )
+                    /
+                   ( (first_point[0]-second_point[0])*(first_point_other[1] - second_point_other[1]) - (first_point[1]-second_point[1])*(first_point_other[0]-second_point_other[0]) );
+
+        return (0.0<=t) && (t<=1.0);
```
Comment (philbucher) [inline kratos/geometries/line_2d_2.h]: no tolerance? Is this intended / how is it done for the other geometries?

**Review by philbucher — CHANGES_REQUESTED** [review]
some comments
and can you please add a test?

File: kratos/geometries/line_2d_2.h (hunk 3 above)
Comment (loumalouomega) [reply kratos/geometries/line_2d_2.h]:
```suggestion
        return (t > -1.0e-12) && (t<(1.0 + 1.0e-12));
```

My suggestion

**Comment by rubenzorrilla** [comment]
Hi @adityaghantasala! Thanks for the implementation. As @philbucher points, I would add some tests for such new feature. When I do this kind of tools I usually add some minor cpp tests considering a simple intersection case, the no-intersection case and a "complex" intersection (e.g. the intersection is parallel to any edge or occurs either in the origin or end points).

**Comment by adityaghantasala** [comment]
test .. on the way .. !

File: kratos/geometries/line_2d_2.h [hunk 4]
```diff
@@ -925,20 +925,44 @@ class Line2D2 : public Geometry<TPointType>
...
+        if (    // If one of the point is inside the box then there is an intersection. If none is inside then we check further.
+                ( (first_point[0] >= rLowPoint[0] && first_point[0] <= rHighPoint[0])
+                    && (first_point[1] >= rLowPoint[1] && first_point[1] <= rHighPoint[1]) ) // IF the first point is inside the box
+                ||
+                  ( (second_point[0] >= rLowPoint[0] && second_point[0] <= rHighPoint[0])
+                    && (second_point[1] >= rLowPoint[1] && second_point[1] <= rHighPoint[1]) ) // IF the second point is inside the box
+            )
+            return true;
+
+        double high_x = rHighPoint[0]; double high_y = rHighPoint[1];
```
Comment (loumalouomega) [inline kratos/geometries/line_2d_2.h]:
```suggestion
        const double high_x = rHighPoint[0]; 
        const double high_y = rHighPoint[1];
```

File: kratos/geometries/line_2d_2.h [hunk 5]
```diff
@@ -925,20 +925,44 @@ class Line2D2 : public Geometry<TPointType>
...
+        double low_x = rLowPoint[0];   double low_y = rLowPoint[1];
```
Comment (loumalouomega) [inline kratos/geometries/line_2d_2.h]:
```suggestion
        const double low_x = rLowPoint[0];   
        const double low_y = rLowPoint[1];
```

File: kratos/geometries/line_2d_2.h [hunk 6]
```diff
@@ -925,20 +925,44 @@ class Line2D2 : public Geometry<TPointType>
...
+
+        double slope = (second_point[1] - first_point[1]) / ( second_point[0] - first_point[0] );
```
Comment (loumalouomega) [inline kratos/geometries/line_2d_2.h]:
```suggestion
        const double denominator = ( second_point[0] - first_point[0] );
        const double numerator = (second_point[1] - first_point[1]);
        const double slope = std::abs(denominator) > tolerance ? std::abs(numerator) > tolerance ? numerator / denominator : 1.0e-12 : 1.0e12;
```

File: kratos/geometries/line_2d_2.h [hunk 7]
```diff
@@ -925,20 +925,44 @@ class Line2D2 : public Geometry<TPointType>
...
+
+        // Intersection with left vertical line of the box that is x = low_x
+        double y_1 = slope*( low_x - first_point[0] ) + first_point[1];
```
Comment (loumalouomega) [inline kratos/geometries/line_2d_2.h]:
```suggestion
      const double y_1 = slope*( low_x - first_point[0] ) + first_point[1];
```

File: kratos/geometries/line_2d_2.h [hunk 8]
```diff
@@ -925,20 +925,44 @@ class Line2D2 : public Geometry<TPointType>
...
+        if(y_1 >= low_y - tolerance && y_1 <= high_y+tolerance) // If y intersection is between two y bounds there is an intersection
+            return true;
+        // Intersection with right vertical line of the box that is x = high_x
+        double y_2 = slope*( high_x - first_point[0] ) + first_point[1];
+        if(y_2 >= low_y - tolerance && y_2 <= high_y+tolerance) // If y intersection is between two y bounds there is an intersection
```
Comment (loumalouomega) [inline kratos/geometries/line_2d_2.h:959]:
```suggestion
        const double y_2 = slope*( high_x - first_point[0] ) + first_point[1];
```

File: kratos/geometries/line_2d_2.h [hunk 9]
```diff
@@ -925,20 +925,44 @@ class Line2D2 : public Geometry<TPointType>
...
+            return true;
+        // Intersection with bottom horizontal line of the box that is y = low_y
+        double x_1 = first_point[0] + ( (low_y - first_point[1]) / slope );
```
Comment (loumalouomega) [inline kratos/geometries/line_2d_2.h]:
```suggestion
        const double x_1 = first_point[0] + ( (low_y - first_point[1]) / slope );
```

File: kratos/geometries/line_2d_2.h [hunk 10]
```diff
@@ -925,20 +925,44 @@ class Line2D2 : public Geometry<TPointType>
...
+        if(x_1 >= low_x-tolerance && x_1 <= high_x+tolerance) // If x intersection is between two x bounds there is an intersection
+            return true;
+        // Intersection with top horizontal line of the box that is y = high_y
+        double x_2 = first_point[0] + ( (high_y - first_point[1]) / slope );
```
Comment (loumalouomega) [inline kratos/geometries/line_2d_2.h]:
```suggestion
       const double x_2 = first_point[0] + ( (high_y - first_point[1]) / slope );
```

File: kratos/geometries/line_2d_2.h (hunk 6 above)
Comment (loumalouomega) [reply kratos/geometries/line_2d_2.h]: This solution cover several corner cases

**Review by loumalouomega — DISMISSED** [review]
I added some suggestions, as adding missing consts or a very relevant one which covers several corner cases (0º and 90º). You can commit directly the suggestions if you agree (is a new feature since yesterday on GitHub)

**Comment by adityaghantasala** [comment]
@loumalouomega what do you mean by the case with 0 and 90 degrees ?

File: kratos/geometries/line_2d_2.h (hunk 8 above)
Comment (adityaghantasala) [reply kratos/geometries/line_2d_2.h:959]: why is this if condition deleted ?

File: kratos/geometries/line_2d_2.h (hunk 6 above)
Comment (adityaghantasala) [reply kratos/geometries/line_2d_2.h]: you mean if the line is horizontal and vertical ?

**Comment by loumalouomega** [comment]
If the denominator is 0 (case of 0º) you will have several corner cases, same for denominator 0 (90º)

File: kratos/geometries/line_2d_2.h (hunk 8 above)
Comment (loumalouomega) [reply kratos/geometries/line_2d_2.h:959]: Huum, probably i did the suggestion wrong, I just wanted to add const

File: kratos/geometries/line_2d_2.h (hunk 6 above)
Comment (loumalouomega) [reply kratos/geometries/line_2d_2.h]: Yes

**Review by loumalouomega — DISMISSED** [review]

**Comment by adityaghantasala** [comment]
@philbucher  concerns addressed ?

File: kratos/geometries/line_2d_2.h [hunk 11]
```diff
@@ -890,6 +890,86 @@ class Line2D2 : public Geometry<TPointType>
...
+        // We get the local points
+        const TPointType& first_point  = BaseType::GetPoint(0); //p1
+        const TPointType& second_point = BaseType::GetPoint(1); //p2
+
+        // We get the other line's points
+        const TPointType& first_point_other  = *rOtherGeometry(0); //p3
+        const TPointType& second_point_other = *rOtherGeometry(1); //p4
+
+        // parametric coordinate of intersection on current line
+        double t = ( (first_point[0]-first_point_other[0])*(first_point_other[1] - second_point_other[1]) - (first_point[1]-first_point_other[1])*(first_point_other[0]-second_point_other[0]) )
```
Comment (philbucher) [inline kratos/geometries/line_2d_2.h]: const

**Review by philbucher — DISMISSED** [review]

**Comment by adityaghantasala** [comment]
sorry for changing again. I added more checks for checking and also one more test.

**Review by loumalouomega — APPROVED** [review]

**Comment by adityaghantasala** [comment]
@loumalouomega thanks .. !

**Pull request merged**